core/src/main/java/cz/o2/proxima/scheme/SchemaDescriptors.java

```diff
@@ -155,147 +283,130 @@ private SchemaDescriptors() {
    */
-  public abstract static class GenericTypeDescriptor<T> implements TypeDescriptor<T> {
-
-    private static final long serialVersionUID = 1L;
-
-    /** Value type */
-    @Getter protected final AttributeValueType type;
-
-    protected GenericTypeDescriptor(AttributeValueType type) {
-      this.type = type;
-    }
-
-    @Override
-    @SuppressWarnings("unchecked")
-    public boolean equals(Object o) {
-      if (this == o) {
-        return true;
-      }
-      if (o instanceof GenericTypeDescriptor) {
-        return type.equals(((GenericTypeDescriptor<?>) o).getType());
-      }
-      return false;
-    }
-
-    @Override
-    public int hashCode() {
-      return Objects.hash(type);
-    }
+  public static <T> StructureTypeDescriptor<T> structures(
+      String name,
+      Map<String, SchemaTypeDescriptor<?>> fields,
+      StructureValueAccessor<T> valueProvider) {
+    return new StructureTypeDescriptor<>(name, fields, valueProvider);
   }
 
   /**
-   * SchemaTypeDescriptor wrapper
+   * Generic type descriptor. Parent class for other types.
    *
    * @param <T> value type
    */
-  public static class SchemaTypeDescriptor<T> implements TypeDescriptor<T> {
+  public abstract static class SchemaTypeDescriptor<T> implements Serializable {
 
-    @Getter final AttributeValueType type;
+    private static final String TYPE_CHECK_ERROR_MESSAGE_TEMPLATE =
+        "Conversion to type %s is not supported. Given type: %s";
 
-    final @Nullable PrimitiveTypeDescriptor<T> primitiveTypeDescriptor;
-    final @Nullable ArrayTypeDescriptor<T> arrayTypeDescriptor;
-    final @Nullable StructureTypeDescriptor<T> structureTypeDescriptor;
+    /** Value type */
+    @Getter protected final AttributeValueType type;
 
-    @SuppressWarnings("squid:S1452")
-    final @Nullable EnumTypeDescriptor<?> enumTypeDescriptor;
-
-    @Builder
-    public SchemaTypeDescriptor(
-        AttributeValueType type,
-        @Nullable PrimitiveTypeDescriptor<T> primitiveTypeDescriptor,
-        @Nullable ArrayTypeDescriptor<T> arrayTypeDescriptor,
-        @Nullable StructureTypeDescriptor<T> structureTypeDescriptor,
-        @Nullable EnumTypeDescriptor<?> enumTypeDescriptor) {
+    protected SchemaTypeDescriptor(AttributeValueType type) {
       this.type = type;
-      this.primitiveTypeDescriptor = primitiveTypeDescriptor;
-      this.arrayTypeDescriptor = arrayTypeDescriptor;
-      this.structureTypeDescriptor = structureTypeDescriptor;
-      this.enumTypeDescriptor = enumTypeDescriptor;
     }
 
-    public PrimitiveTypeDescriptor<T> getPrimitiveTypeDescriptor() {
-      Preconditions.checkState(
-          isPrimitiveType(), "SchemaTypeDescriptor is not primitive type. Type: " + getType());
-      return primitiveTypeDescriptor;
+    /**
+     * Return {@code true} if type is primitive.
+     *
+     * @return boolean
+     */
+    public boolean isPrimitiveType() {
+      return this instanceof PrimitiveTypeDescriptor;
```

Comment:
Yuck ... this is the canonical example of how *not* to use instanceof :)
Please make it abstract and let the subclasses override this.